compiler/rustc_resolve/src/late.rs

```diff
@@ -4224,13 +4224,16 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             .entry((Symbol::intern(path_str), ns))
             .or_insert_with_key(|(path, ns)| {
                 let res = self.r.resolve_rustdoc_path(path.as_str(), *ns, self.parent_scope);
-                if let Some(res) = res
-                    && let Some(def_id) = res.opt_def_id()
-                    && !def_id.is_local()
-                    && self.r.tcx.sess.crate_types().contains(&CrateType::ProcMacro)
-                    && matches!(self.r.tcx.sess.opts.resolve_doc_links, ResolveDocLinks::ExportedMetadata) {
-                    // Encoding foreign def ids in proc macro crate metadata will ICE.
-                    return None;
+                if let Some(res) = res {
+                    if matches!(res, Res::NonMacroAttr(_) | Res::ToolMod) {
```

Comment:
I don't understand how this fixes the issue.
Why does rustdoc crash if this (correct) resolution ends up in the table?

Reply:
The logic here is correct - there is indeed a resolution, and the function returns it, the issue is somewhere else.

Reply:
The resolution is wrong but that's another topic. If the item is resolved, then `rustfmt::skip` is saved as is. Problem is that rustdoc will then try to resolve `rustfmt`, which isn't stored. By rejecting these `Res`, we will save both `rustfmt` and `rustfmt::skip` (look for `while let Some(idx) = path.rfind("::") {` in `rustc_resolve/src/late.rs`).

Reply:
And no, these `Res` are invalid. They don't refer to something that actually exist as far as I understand. But maybe I missed something?

Reply:
>Problem is that rustdoc will then try to resolve `rustfmt`, which isn't stored.

Why does rustdoc try to do that?
AFAIR, it only did that for unsuccessful resolutions, but this is a successful resolution (in `fn resolve_doc_links` `any_resolved` becomes true so we don't resolve prefixes).

Reply:
Because rustdoc is only looking for `Res::Def` and `Res::Primitive`. If we can't link to it, we discard it.

Reply:
>If we can't link to it, we discard it.

In this case rustdoc should probably say "your path points to a <tool attribute/built-in attribute/etc>, I cannot create a link to it" instead of treating it as entirely unresolved and starting resolving prefixes.

Reply:
No because we want to provide suggestions. For example if it's an intra doc link to an enum variant, we want to go through the enum variants to tell "hey, you did a typo, here are the variants available".

Reply:
Well, if the whole path resolves to a tool attribute, then all the prefixes will resolve to tool modules, you won't get any suggestions out of them.

Reply:
Yes but I'd rather not add more specific cases handling on rustdoc side when we can simply handle with very small changes in rustc directly.

Reply:
The least hacky way to fix this in rustc would be something like https://github.com/petrochenkov/rust/commit/721c695d9a4cec7ce548da2f7cccdecc4ef6cdaa